### ISSUE TITLE
Resource transforms: alter URLs before requesting

### ISSUE THREAD

**Issue by kkaefer** [mapbox/mapbox-gl-js#4740]
In Native, [we have added "Resource transforms"](https://github.com/mapbox/mapbox-gl-native/pull/8084), which allows you to supply a function that is called every time a URL is requested. That function can perform arbitrary operations on the URL string, and the returned URL will be requested. We should implement this in GL JS for API parity.

**Comment by Scarysize** [comment]
Will this feature make it possible to have relative URLs for certain properties in the map style JSON? E.g.:

```js
{
  ...
  "spritesheet": "my/local/spritesheet"
}
```

A callback could resolve this relative url and return an absolute one.

**Comment by asheemmamoowala** [comment]
@Scarysize - the changes proposed in #4763 do not explicitly allow pure relative URLs in the `style.json`. 
Instead, you can provide a custom protocol and resolve that in the transform callback:
```
{
...
"sprite": "app-local://my/local/spritesheet"
}
```

**Comment by Scarysize** [comment]
That was my intention. Thanks for confirming that 🙌

**Comment by volkergraf** [comment]
Will this kinda make the following Possible:

To load "local"-Tiles like 
```
"sources": {
                "openmaptiles": {
                    "type": "vector",
                    "tiles":  [location.href + "/../germany/{z}/{x}/{y}.pbf" ],
                    "name": "OpenMapTiles",
                    "format": "pbf",
                    "basename": "germany.mbtiles",
```

on iOS, I have to transform the FileURL from e.g. **location.href + "/../germany/14/1111/{2222.pbf"**
to the corresponding blob-id (_blob:file:///c415dbb6-3d57-4354-83c0-a33c0b6ddb4f_) on iOs using the function  _window.resolveLocalFileSystemURL_ provided by the **cordova-plugin-file**-Plugin. 

Thanx for any Help

**Comment by asheemmamoowala** [comment]
@volkergraf I believe you are asking for #4917. 

#5021 will allow transforming the URL as you've mentioned above. Not being familiar with Cordova,  I do not know if a blob url from Cordova,  returned from `window.resolveLocalFuleSystemURL`, can be retrieved using an `XMLHttpRequest`.

**Comment by volkergraf** [comment]
@asheemmamoowala Well THIS I can definively can confirm (using the cordova-file-plugin and the local-URl-Fix).. to test it I created such an URL form an Object like

  **./germany/14/8601/5648.pbf -> blob:file:///55e863d2-00d0-4712-b7ad-ab2792505e14**
```
"sources": {
                "openmaptiles": {
                    "type": "vector",
                    "tiles":  [**"blob:file:///55e863d2-00d0-4712-b7ad-ab2792505e14"** ],
```

And I see the tile .. but well everywhere =) .. but at least I see that it's beeing loaded =)

**Comment by asheemmamoowala** [comment]
Fixed in #5021
